mirrord/agent/src/cgroup.rs

```diff
@@ -0,0 +1,256 @@
+//! Logic for pausing using cgroup directly
+//! Pause requires privileged - assumes ephemeral (hardcoded pid 1)
+use std::{
+    fs::OpenOptions,
+    io::{BufRead, BufReader, Write},
+    path::{Path, PathBuf},
+};
+
+use enum_dispatch::enum_dispatch;
+use nix::mount::{mount, MsFlags};
+use tracing::{trace, warn};
+
+use crate::{
+    error::{AgentError, Result},
+    namespace::{set_namespace, NamespaceType},
+};
+
+const CGROUP_MOUNT_PATH: &str = "/mirrord_cgroup";
+const CGROUP_SUBGROUP_MOUNT_PATH: &str = "/mirrord_cgroup/subgroup";
+const CGROUPV2_PROCS_FILE: &str = "cgroup.procs";
+
+/// Trait for objects that can be paused
+#[enum_dispatch]
+pub(crate) trait CgroupFreeze {
+    fn pause(&self) -> Result<()>;
+    fn unpause(&self) -> Result<()>;
+}
+
+#[derive(Debug)]
+pub(crate) struct CgroupV1 {
+    /// Path to the the process' cgroup - usually /kubepods/besteffort/ID
+    cgroup_path: PathBuf,
+}
+
+impl CgroupV1 {
+    #[tracing::instrument(level = "trace", ret)]
+    pub(crate) fn new() -> Result<Self> {
+        let file = std::fs::File::open("/proc/1/cgroup")
+            .map_err(|_| AgentError::PauseFailedCgroup("opening proc cgroup failed".to_string()))?;
+        let reader = std::io::BufReader::new(file).lines();
+        for line in reader {
+            let line = line?;
+            let mut line_iter = line.split(':');
+            // we don't care about the number, prefer the string for comparison
+            line_iter.next().ok_or(AgentError::PauseFailedCgroup(
+                "malformed ID cgroup v1 file".to_string(),
+            ))?;
+            let cgroup_type = line_iter.next().ok_or_else(|| {
+                AgentError::PauseFailedCgroup("malformed cgroup type cgroup v1 file".to_string())
+            })?;
+            if cgroup_type == "freezer" {
```

Comment:
```suggestion
let cgroup_type = line.split(':').find(|value| value == "freezer").ok_or_else(...)?;
```

Can't this be reduced to something like this?

Reply:
I prefer a more explicit parsing so if the format isn't what we expect we error, and we know beter what's going on..

Reply:
Ouch, but this kinda adds a lot of noise, because we don't really care about all these in-between steps (seeing that they all pretty much return the same error), the goal here is to either get the path with "freeze" or fail.

The extra steps feels likes debug code, like we're doing a manual `inspect` trying to find what's going wrong.